nm ovs: Enforce compliance of OVS port VLAN schema

cathay4t opened the pull request:
The OVS port vlan schema is:

```yml
  - name: ovs-br0
    type: ovs-bridge
    state: up
    bridge:
      port:
        - name: ovs0
          vlan:
            mode: access
            tag: 2
```

Add two test cases:
    * Example test against `examples/ovsbridge_vlan_port.yml`.
    * Integration test: `test_ovs_vlan_access_tag`

tyll commented:
This will conflict with #763, please coordinate the changes.

cathay4t commented:
> This will conflict with #763, please coordinate the changes.

OK. I will wait #763 been merged and rebase from there.

tyll commented:
> > This will conflict with #763, please coordinate the changes.
> 
> OK. I will wait #763 been merged and rebase from there.

It is merged, please rebase.

cathay4t commented:
Rebased.

EdDev commented on tests/integration/testlib/ovslib.py:
```diff
@@ -91,6 +91,16 @@ def _get_ports(self):
+        ):
+            if (
+                port_option[OVSBridge.Port.NAME]
+                == new_option[OVSBridge.Port.NAME]
```
Can you make this an explicit argument so it will be clear and explicit that this is the port to change?

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -173,15 +173,16 @@ def _get_bridge_port_info(port_profile, devices_info):
         number_of_interfaces = len(port_slave_names)
         if number_of_interfaces == 1:
             port_info[OB.Port.NAME] = port_slave_names[0]
+            if vlan_mode:
```
The vlan mode as seen from NM/ovsdb is not access/trunk, it is a bunch of other values. These need to be "translated".
Specifically:
access => access
trunk => trunk
native-untagged => trunk
// The below are rare, and will require an additional flag probably if requested.
// We should not support them at the moment.
native-tagged => trunk
dot1q-tunnel => trunk

cathay4t replied:
You want me to do that in this PR?

cathay4t replied:
If we don't differentiate the `trunk` and `native-untagged`, when user apply `trunk` mode, how should we know which one is user requested.

I intend to keep it as it was.

EdDev replied:
I prefer the mode not to include this subset cases as OVSDB exposes but to use flags.
The `native-untagged` is covered through `enable-native` flag and `mode` is accepting only `trunk` and `access` in the schema (in sync with the Linux Bridge VLAN filter as well).

cathay4t replied:
What will happen when OVS introduce a new mode?
We raise exception?

EdDev replied:
Good point. We can add to the schema an `unknown` option.
But IMO, any new option may be classified as trunk and we can print a a warning/error that it is not a supported mode in nmstate.

cathay4t commented on libnmstate/nm/ovs.py:
```diff
@@ -85,11 +85,11 @@ def create_port_setting(port_state):
-            port_setting.props.vlan_mode = option_value
+    vlan_state = port_state.get(OB.Port.VLAN_SUBTREE, {})
+    if OB.Port.Vlan.MODE in vlan_state:
+        port_setting.props.vlan_mode = vlan_state[OB.Port.Vlan.MODE]
```
Py 3.6 is the minimum support python version.

EdDev replied:
I mean that `OB.Port.Vlan.MODE in vlan_state` and `vlan_state[OB.Port.Vlan.MODE]` perform a lookup when once should be suffice. (fetch that data and then query it)

cathay4t commented:
New patch uploaded.

Changes since previous patch:
 * The VLAN information is not limited to single OVS port now.
 * The test code `testlib/ovs.py set_port_option()` now require port name explicitly.

cathay4t commented:
New patch uploaded.

Added `OvsBridge.OB.Port.Vlan.Mode.UNKNOWN` to schema for unsupported VLAN modes.

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -175,10 +184,17 @@ def _get_bridge_port_info(port_profile, devices_info):
+                    f"OVS Port VLAN mode '{vlan_mode}' is not supported yet"
+                )
+            port_info[OB.Port.VLAN_SUBTREE] = {
+                OB.Port.Vlan.MODE: nmstate_vlan_mode,
```
This will be set even for unsupported modes, so it will show here `None`. Is this your intention? Maybe set `UNKNOWN`.

cathay4t replied:
My fault. Will fix it.

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -175,10 +184,17 @@ def _get_bridge_port_info(port_profile, devices_info):
+                )
+            port_info[OB.Port.VLAN_SUBTREE] = {
+                OB.Port.Vlan.MODE: nmstate_vlan_mode,
+                OB.Port.Vlan.TAG: port_setting.props.tag,
```
What if there is nothing set? Maybe normalize to `0`? Or just do not add it.

cathay4t replied:
The code of NetworkManager `nm_setting_ovs_port_get_tag()` said if not set, return 0.
Changed to use `get_tag()` function.

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -175,10 +185,17 @@ def _get_bridge_port_info(port_profile, devices_info):
-            port_info["access-tag"] = port_setting.props.tag
 
+        if vlan_mode:
+            nmstate_vlan_mode = NM_OVS_VLAN_MODE_MAP.get(vlan_mode)
```
The default should be `OB.Port.Vlan.Mode.UNKNOWN` in case a new mode appears.

EdDev commented on libnmstate/nm/ovs.py:
```diff
@@ -175,10 +185,17 @@ def _get_bridge_port_info(port_profile, devices_info):
+                )
+            port_info[OB.Port.VLAN_SUBTREE] = {
+                OB.Port.Vlan.MODE: nmstate_vlan_mode,
+                OB.Port.Vlan.TAG: port_setting.get_tag(),
```
`port_setting.get_tag()` can return None, no? Maybe add an `or 0`?

cathay4t replied:
The document and NetworkManager confirmed, it will return a integer and 0 for missing.
